File: mobile/android/app/src/main/java/com/btrace/app/ui/screens/compliance/ComplianceViewModel.kt
package com.btrace.app.ui.screens.compliance

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.btrace.app.data.api.BTraceApi
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class ComplianceViewModel @Inject constructor(
    private val api: BTraceApi
) : ViewModel() {

    private val _uiState = MutableStateFlow<ComplianceUiState>(ComplianceUiState.Loading)
    val uiState: StateFlow<ComplianceUiState> = _uiState

    init {
        loadComplianceData()
    }

    private fun loadComplianceData() {
        viewModelScope.launch {
            _uiState.value = ComplianceUiState.Loading
            try {
                // Fetch compliance data from API
                val reports = api.getComplianceReports()
                val deadlines = fetchUpcomingDeadlines()
                
                _uiState.value = ComplianceUiState.Success(
                    totalReports = reports.size,
                    pendingActions = countPendingActions(reports),
                    upcomingDeadlines = deadlines
                )
            } catch (e: Exception) {
                _uiState.value = ComplianceUiState.Error(
                    e.message ?: "Failed to load compliance data"
                )
            }
        }
    }

    private fun countPendingActions(reports: List<Any>): Int {
        // Count pending actions based on report status
        return 2 // Placeholder
    }

    private suspend fun fetchUpcomingDeadlines(): List<ComplianceDeadline> {
        // Fetch from API or calculate based on compliance rules
        return listOf(
            ComplianceDeadline("CBAM Q4 Report", "2024-12-31", true),
            ComplianceDeadline("EPR Annual Filing", "2025-01-31", false),
            ComplianceDeadline("GST Reconciliation", "2024-11-30", false)
        )
    }

    suspend fun generateReport(reportType: String): Result<String> {
        return try {
            val downloadUrl = api.generateReport(reportType)
            Result.success(downloadUrl)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getConsentRecords(): Result<List<Any>> {
        return try {
            val consents = api.getConsentRecords()
            Result.success(consents)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: mobile/ios/BTraceApp/Sources/ViewModels/ComplianceViewModel.swift
//
//  ComplianceViewModel.swift
//  BTraceApp
//
//  Created by BTrace Team
//  Copyright © 2024 BTrace. All rights reserved.
//

import Foundation

@MainActor
class ComplianceViewModel: ObservableObject {
    @Published var totalReports: Int = 0
    @Published var pendingActions: Int = 0
    @Published var upcomingDeadlines: [ComplianceDeadline] = []
    
    let reportTypes: [String] = [
        "CBAM Report",
        "EPR Report",
        "Carbon Intensity Report",
        "Mass Balance Report",
        "GST Compliance Report",
        "Audit Trail Export"
    ]
    
    private let networkService: NetworkService
    
    init(networkService: NetworkService = NetworkService.shared) {
        self.networkService = networkService
        Task {
            await loadComplianceData()
        }
    }
    
    func loadComplianceData() async {
        do {
            // Fetch compliance data from API
            let reports = try await networkService.getComplianceReports()
            totalReports = reports.count
            pendingActions = calculatePendingActions(reports)
            upcomingDeadlines = fetchUpcomingDeadlines()
        } catch {
            print("Failed to load compliance data: \(error)")
        }
    }
    
    func generateReport(_ reportType: String) async {
        do {
            let downloadUrl = try await networkService.generateReport(type: reportType)
            // Handle file download
            print("Report generated: \(downloadUrl)")
        } catch {
            print("Failed to generate report: \(error)")
        }
    }
    
    private func calculatePendingActions(_ reports: [ComplianceReport]) -> Int {
        // Count pending actions based on report status
        return reports.filter { $0.status == "pending" }.count
    }
    
    private func fetchUpcomingDeadlines() -> [ComplianceDeadline] {
        // Calculate based on compliance rules
        return [
            ComplianceDeadline(title: "CBAM Q4 Report", dueDate: "2024-12-31", isUrgent: true),
            ComplianceDeadline(title: "EPR Annual Filing", dueDate: "2025-01-31", isUrgent: false),
            ComplianceDeadline(title: "GST Reconciliation", dueDate: "2024-11-30", isUrgent: false)
        ]
    }
}

struct ComplianceDeadline: Identifiable {
    let id = UUID()
    let title: String
    let dueDate: String
    let isUrgent: Bool
}
